mcs/class/corlib/Mono/SafeGPtrArrayHandle.cs

```diff
@@ -0,0 +1,42 @@
+//
+// Safe handle class for Mono.RuntimeGPtrArrayHandle
+//
+// Authors:
+//   Aleksey Kliger <aleksey@xamarin.com>
+//   Rodrigo Kumpera <kumpera@xamarin.com>
+//
+// Copyright 2016 Dot net foundation.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Mono {
+	internal sealed class SafeGPtrArrayHandle : IDisposable {
+		RuntimeGPtrArrayHandle handle;
+		bool freeSeg;
+
+		internal SafeGPtrArrayHandle (IntPtr ptr, bool freeSeg)
+		{
+			handle = new RuntimeGPtrArrayHandle (ptr);
+			this.freeSeg = freeSeg;
+		}
+
+		public void Dispose ()
+		{
+			RuntimeGPtrArrayHandle.DestroyAndFree (ref handle, freeSeg);
+			GC.SuppressFinalize (this);
```

Comment:
No need to suppress the finalizer of a class without a finalizer.
Maybe it's missing a finalizer.

Another thing. Once you dispose, you must change the handle to disposed to ensure we'll fail cleanly instead of accessing a dangling pointer.